### PR TITLE
Project, app, workspace, and label scoping for Runner Config

### PR DESCRIPTION
**Pull request by mitchellh**
This hooks up the deep core work of #2169 to one aspect of real functionality: runners now properly look up config for their scope. So if you set a config that is for runners running jobs for the "foo" project, on-demand runners will now properly get that configuration.

Note this only works for on-demand runners. This was a limitation we expected from the beginning.

## The Journey

We're not quite done, I'm continuing to keep my PRs small and focused. Our journey so far:

1. [x] #2169 set up our core internal data model. 
2. [x] **This PR** hooks it up to runners so once set, runners actually see it.
3. [ ] Future PR will add CLI work so that users can actually set this.
4. [ ] Future PR (not by me) will add UI work so users can set this in the UI.
5. [ ] Future PR: docs on the new config scoping

### CODE REVIEW

File: internal/server/singleprocess/service_runner.go [hunk 1]
```diff
@@ -110,22 +109,65 @@ func (s *service) RunnerConfig(
...
+			return status.Errorf(codes.FailedPrecondition,
+				"no pending job for this on-demand runner. A pending job "+
+					"must be registered prior to registering the runner.")
+		}
+
+		// Nack the job cause we're not going to take it ourselves,
+		// the runner will later with JobStream.
+		if _, err := s.state.JobAck(sjob.Id, false); err != nil {
+			return err
+		}
```
Comment (evanphx) [inline internal/server/singleprocess/service_runner.go:136]: Should we have a `JobPeekForRunner` that doesn't perform the assignment?

**Review by evanphx — CHANGES_REQUESTED** [review]
Initial review, mostly wondering about having an API to peek at the job queue.

File: internal/server/singleprocess/service_runner.go (hunk 1 above)
Comment (evanphx) [reply internal/server/singleprocess/service_runner.go:136]: Didn't mean this to be a change request, just a comment.

Comment (mitchellh) [reply internal/server/singleprocess/service_runner.go:136]: Yeah, this is probably a better idea.

Comment (mitchellh) [reply internal/server/singleprocess/service_runner.go]: Done

**Review by evanphx — APPROVED** [review]

**Pull request merged**